src/sentry/flags/endpoints/hooks.py

```diff
@@ -0,0 +1,161 @@
+from sentry.models.organization import Organization
+from sentry.utils.sdk import bind_organization_context
+
+"""HTTP endpoint.
+
+This endpoint accepts only organization authorization tokens. I've made the conscious
+decision to exclude all other forms of authentication. We don't want users accidentally
+writing logs or leaked DSNs generating invalid log entries. An organization token is
+secret and reasonably restricted and so makes sense for this use case where we have
+inter-provider communication.
+
+This endpoint allows writes if any write-level "org" permission was provided.
+"""
+
+
+class OrganizationFlagHookPermission(OrganizationPermission):
+    scope_map = {
+        "POST": ["org:ci"],
+    }
+
+
+@region_silo_endpoint
+class OrganizationFlagsHooksEndpoint(Endpoint):
+    authentication_classes = (OrgAuthTokenAuthentication,)
+    owner = ApiOwner.REPLAY
+    permission_classes = (OrganizationFlagHookPermission,)
+    publish_status = {
+        "POST": ApiPublishStatus.PRIVATE,
+    }
+
+    def convert_args(
+        self,
+        request: Request,
+        organization_id_or_slug: int | str,
+        *args,
+        **kwargs,
+    ):
+        try:
+            if isinstance(organization_id_or_slug, int):
+                organization = Organization.objects.get_from_cache(id=organization_id_or_slug)
+            else:
+                organization = Organization.objects.get_from_cache(slug=organization_id_or_slug)
+        except Organization.DoesNotExist:
+            raise ResourceDoesNotExist
+
+        self.check_object_permissions(request, organization)
+        bind_organization_context(organization)
+
+        kwargs["organization"] = organization
+        return args, kwargs
+
+    def post(self, request: Request, organization: Organization, provider: str) -> Response:
+        try:
+            rows_data = handle_provider_event(provider, request.data, organization.id)
+            FlagAuditLogModel.objects.bulk_create(FlagAuditLogModel(**row) for row in rows_data)
+            return Response(status=200)
+        except InvalidProvider:
+            raise ResourceDoesNotExist
+        except DeserializationError as exc:
+            return Response(exc.errors, status=400)
+
+
+"""Provider definitions.
+
+Provider definitions are pure functions. They accept data and return data. Providers do not
+initiate any IO operations. Instead they return commands in the form of the return type or
+an exception. These commands inform the caller (the endpoint defintion) what IO must be
+emitted to satisfy the request. This is done primarily to improve testability and test
+performance but secondarily to allow easy extension of the endpoint without knowledge of
+the underlying systems.
+"""
+
+
+class FlagAuditLogRow(TypedDict):
+    """A complete flag audit log row instance."""
+
+    action: int
+    flag: str
+    modified_at: datetime.datetime
+    modified_by: str
+    modified_by_type: int
+    organization_id: int
```

Comment:
the `organization_id` field is missing from the blueprint

Reply:
It's not exposed.  The organization_id is stored on the row but we never explicitly return it.

Reply:
oh gotcha, missed that!